### PR TITLE
[ConstraintSystem] Use `missing conformance` fix to diagnose contextual failures

### PR DESCRIPTION
**Pull request by xedin**
Extend use of `missing protocol conformance` fix to cover contextual
failures, such as:

- Assignment mismatches, where destination requires source to conform
   to certain protocol (or protocol composition);
- Incorrect returns where returned type doesn't conform to the
   protocol specified in the signature.

### CODE REVIEW

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -2040,6 +2040,19 @@ bool ConstraintSystem::repairFailures(
     return false;
   };
 
+  auto repairByAddingConformance = [&](Type lhs, Type rhs) -> bool {
+    if (lhs->isTypeVariableOrMember())
+      return false;
+
+    if (rhs->isAny() ||
+        !(rhs->is<ProtocolType>() || rhs->is<ProtocolCompositionType>()))
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: !rhs->isExistentialType()

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -2040,6 +2040,19 @@ bool ConstraintSystem::repairFailures(
     return false;
   };
 
+  auto repairByAddingConformance = [&](Type lhs, Type rhs) -> bool {
+    if (lhs->isTypeVariableOrMember())
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: What happens if you don’t bail out in this case?

File: test/Generics/conditional_conformances_literals.swift [hunk 3]
```diff
@@ -51,7 +51,7 @@ func dictionarySameType() {
 
     let _: SameType = [0 : works]
     let _: SameType = [0 : fails]
-    // expected-error@-1 {{contextual type 'SameType' cannot be used with dictionary literal}}
+    // expected-error@-1 {{value of type '[Int : Fails]' does not conform to specified type 'SameType'}}
```
Comment (slavapestov) [inline test/Generics/conditional_conformances_literals.swift]: Nice!

**Comment by xedin** [comment]
@slavapestov This is still a draft, I'm trying to work out couple of problems related to conditional requirements and add some more tests.

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Yeah, I think it makes sense to test that too.

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: I think it's going to clash with Existential restriction, but I'm still working out all of the logic here...

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Sema/CSSimplify.cpp]: Sorry, I mean literally isExistentialType() is equivalent to checking for ProtocolType and ProtocolCompositionType :)

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Ah, thanks! I am going to change that in CSDiag as well then!

**Comment by xedin** [comment]
@slavapestov I have moved fix to `matchExistentialTypes`, which seems to be a much better place for it, and allows to diagnose failures related to protocol compositions more precisely by pointing out exactly which conformances are missing. Please take a look.

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/CSDiag.cpp [hunk 4]
```diff
@@ -2316,10 +2303,13 @@ bool FailureDiagnosis::diagnoseContextualConversionError(
 
   // When complaining about conversion to a protocol type, complain about
   // conformance instead of "conversion".
-  if (contextualType->is<ProtocolType>() ||
-      contextualType->is<ProtocolCompositionType>())
-    diagID = diagIDProtocol;
-  
+  if (contextualType->isExistentialType()) {
```
Comment (DougGregor) [inline lib/Sema/CSDiag.cpp]: Should we do the same for opaque result types and/or AnyHashable? They have the same kinds of conformance-related issues.

**Review by DougGregor — APPROVED** [review]

File: lib/Sema/CSDiag.cpp (hunk 4 above)
Comment (xedin) [reply lib/Sema/CSDiag.cpp]: Yeah, those would already be diagnose by new diagnostic framework, I have a couple of examples added to the suite in my previous PR :)

**Comment by davezarzycki** [comment]
This broke one test on my Linux machine (Red Hat Fedora 30). Was this expected?
```
: 'RUN: at line 1';   /home/dave/s/u/t/bin/swift -frontend -target x86_64-unknown-linux-gnu  -module-cache-path '/home/dave/s/u/t/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -swift-version 4   -typecheck -verify -disable-objc-attr-requires-foundation-module /home/dave/s/u/swift/test/stmt/foreach.swift
--
Exit Code: 1

Command Output (stderr):
--
/home/dave/s/u/swift/test/stmt/foreach.swift:64:58: error: incorrect message found
  for f : ElementProtocol in gir { } // expected-error {{sequence element type 'Int' does not conform to expected type 'ElementProtocol'}}
                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                         type 'Int' does not conform to protocol 'ElementProtocol'
```

**Comment by xedin** [comment]
Yeah, I should have re-ran the tests...

**Pull request merged**